### ISSUE TITLE
Copy on Write/reflink for supported filesystems

### ISSUE THREAD

**Issue by rubdos** [beetbox/beets#2642]
For the uninitiated, [this article explains what reflinks are](https://lwn.net/Articles/331808/).

### Problem

There are use cases where the user wants the data which they import using `beets im` to be untouched, unmoved. Currently, the only option is `copy`, which obviously makes a copy of the original to-be-imported file. This (obviously) effectively duplicates the total amount of data for the end-user, while actually beets (not considering transcoding options) only changes a few kB at most.

It would thus be beneficial to deduplicate said data, preferable *without* copying the file at all. Currently, `copy` followed by a (very CPU and disk intensive) deduplication action saves same space, but using reflinks on supported filesystems would increase both performance of beets (achieving the same speed as a hard link), while keeping the benefits from copy.

### Design decisions

I propose a tristate configuration option `reflink`, such as `cp --reflink` provides:

- no: do not use reflink (*default*)
- always: force reflink, failing when unable to use reflink
- auto: try using reflink, but fallback onto `copy` when fails

---

*I am currently working on this, expect a WIP PR soon.* Feedback before is obviously welcome :-)

**Comment by sampsyo** [comment]
Sounds great. This has actually come up before on the mailing list: https://groups.google.com/forum/#!topic/beets-users/QpLnNCgRkC8

The main barrier seems to be accessing the necessary syscalls from Python. Also, for what it's worth, it would also be great to support the brand-new macOS/APFS version of the same, which is known as clonefile(2) and accessible as `cp -c`.

I'd gently argue that `auto` should be the default. It would be nice for users to get the performance benefits "for free" when their system supports it—especially if we can be robust enough, such as by detecting cross-filesystem copies and degrading gracefully, that there's no real downside.

**Comment by rubdos** [comment]
> The main barrier seems to be accessing the necessary syscalls from Python. Also, for what it's worth, it would also be great to support the brand-new macOS/APFS version of the same, which is accessible as cp -c. Here's some example C code for using the API on macOS.

I'm creating an [external library for reflink](https://gitlab.com/rubdos/pyreflink), such that beets doesn't have to do all the OS dependent stuff, and maximizing reuse.

I do not have a macOS system to test on, but I will gratefully accept patches for pyreflink for macOS :-)

> I'd gently argue that auto should be the default. It would be nice for users to get the performance benefits "for free" when their system supports it—especially if we can be robust enough, such as by detecting cross-filesystem copies and degrading gracefully, that there's no real downside.

The detection of cross-filesystem is implemented in pyreflink (through the result of the system call), but may be it is better to also do it in beets itself, avoiding the syscall at all? Not sure there; it might be useful to postpone that to a separate issue.

---

I currently have some troubles doing decent testing of the pyreflink library on GitLab, but the tests run on my local system, so I'm at a point that I can call it from beets now.

I'll keep you up to date :-)

**Comment by rubdos** [comment]
> I'd gently argue that auto should be the default. It would be nice for users to get the performance benefits "for free" when their system supports it

So, move away from copy being the default?

**Comment by sampsyo** [comment]
Oh wow; that's great! It seems like the Python community could really benefit from a library for this. I'll look into the Apple APIs. It may be possible to do it with ctypes, which would avoid the need for native code.

> So, move away from copy being the default?

I was actually trying to see a CoW as just as transparent optimization on "copy." That is, the user can just think of it as copying but faster. On further reflection, the right thing to do is probably to have a separate option and start out with it off by default—we can turn it on once we're confident we've gotten things right.

**Comment by rubdos** [comment]
> Oh wow; that's great! It seems like the Python community could really benefit from a library for this. I'll look into the Apple APIs. It may be possible to do it with ctypes, which would avoid the need for native code.

Please refrain from using ctypes, and use cffi instead. It makes it possible to run under third party Python interpreters. I already used cffi for the Linux driver, so please stick to that. It's not difficult at all, and you don't even need native code (although you can use it, as I did).

From the looks of the Apple API, it seems like it's even less code than for Linux, and having native code would only be a possible performance optimization.

> I was actually trying to see a CoW as just as transparent optimization on "copy." That is, the user can just think of it as copying but faster. On further reflection, the right thing to do is probably to have a separate option and start out with it off by default—we can turn it on once we're confident we've gotten things right.

Agreed. Shipping at all is blocked on [this permissions issue](https://gitlab.com/rubdos/pyreflink/issues/3).

**Comment by axelsimon** [comment]
Is there any chance of seeing this getting completed? :slightly_smiling_face: Looks like you closed the permissions issue you mentioned a few days later, @rubdos. Did you manage to unblock the issue?

**Comment by rubdos** [comment]
Merge request is at https://github.com/beetbox/beets/pull/3088, basically waiting on a bit of discussion on the exception handling.
